wallet, test: Ancient Wallet Migration from v0.14.3 (no-HD and Single Chain)

w0xlt opened the pull request:
This PR adds test coverage for migrating legacy Bitcoin Core wallets from v0.14.3 (released in 2017) to the descriptor wallet format. The test validates that users can safely upgrade their wallets while preserving all funds, transaction history, and addresses.

This test was originally developed on top of #32977, as it was requested in reviews.
However, since it also increases test coverage, it can be merged independently.

The test covers two wallet migration scenarios:

* Non-HD Wallet Migration - Tests migration of non-HD wallets (created with `-usehd=0`)
* Single Chain HD Wallet Migration - Tests migration of HD wallets from v0.14.3 (`VERSION_HD_BASE`)

The node v0.14.3 cannot be synced using the normal test framework helpers because it does not have the `syncwithvalidationinterfacequeue` RPC, so the test uses `dumb_sync_blocks` to submit blocks from the ancient node to the modern node before migrating the wallet.

Each scenario uses its own dedicated old/new node pair, which keeps the setup isolated and makes this testing best managed in a separate file rather than in the existing migration test files.

On the Windows cross-built CI job, this test is excluded from the main functional test runner and re-run sequentially in an ASCII-only tmpdir, because the v0.14.3 binary cannot handle non-ASCII characters in the temporary directory path.

w0xlt committed:
test: support v0.14.x in dumb_sync_blocks

The getblock RPC used a boolean verbose argument before v0.15. Use the
legacy named argument so the helper can synchronize blocks from v0.14.x
nodes while remaining compatible with current nodes.

w0xlt committed:
test: cover v0.14.3 wallet migration

Test migratewallet on v0.14.3 non-HD and single-chain HD wallets in both
unencrypted and encrypted configurations.

Verify balances, transaction history, address ownership, descriptor
structure, encryption enforcement, backup creation, and the absence of
rescans or unexpected auxiliary wallets.

Run the test with an ASCII-only temporary directory in the Windows
cross-build job because the v0.14.3 binary cannot handle the Unicode
runner path.

pablomartin4btc commented on test/functional/wallet_ancient_migration.py:
```diff
@@ -0,0 +1,292 @@
+
+    def setup_network(self):
+        # Don't setup nodes here, we'll do it per test
+        self.nodes = []
```
nit: I think if you call `cleanup_nodes()` from within `setup_network()`, you don't need to call it from `migrate_wallet()`...
```suggestion
        self.nodes = []
        cleanup_nodes()
```

pablomartin4btc reviewed (commented):
Concept ACK

Left a couple of suggestions.

w0xlt commented on test/functional/test_runner.py:
```diff
@@ -428,7 +429,12 @@ def main():
     # Create base test directory
-    tmpdir = "%s/test_runner_₿_🏃_%s" % (args.tmpdirprefix, datetime.datetime.now().strftime("%Y%m%d_%H%M%S"))
+    # Special case for tests using old binaries (e.g. version v0.14.3) that don't handle Unicode on Windows
+    has_migration_tests = any('wallet_ancient_migration' in script for script in BASE_SCRIPTS)
```
Without this, the `Windows, test cross-built` CI task will fail because it can't handle non-ASCII characters in `-datadir`.

`feature_unsupported_utxo_db.py` doesn't copy the directory from one node to another. This test needs to do this because version 0.14.3 doesn't have the `syncwithvalidationinterfacequeue` RPC, so it can't be synchronized.

pablomartin4btc replied:
Yeah, it's already well explained in the PR description. Then perhaps the variable name could be `supports_unicode_datadir`/ (`requires_ascii_datadir`) or something like that.
```suggestion
    requires_ascii_datadir = any('wallet_ancient_migration' in script for script in BASE_SCRIPTS)
    tmpdirprefix = "%s/test_runner_₿_🏃_%s"
    if platform.system() == 'Windows' and requires_ascii_datadir:
        tmpdir = "%s/test_runner_btc_run_%s"
    tmpdir = tmpdirprefix % (args.tmpdirprefix, datetime.datetime.now().strftime("%Y%m%d_%H%M%S"))
```

Wouldn't this affect the datadir for all the tests running on Windows platform?

w0xlt replied:
Thanks for the suggestion. It's much clearer. 
Yes, it affects all tests run on the Windows platform. As far as I know, there's no way to filter by this task specifically.

fanquake commented:
@achow101 can you leave a conceptual review?

furszy reviewed (commented):
> The node v0.14.3 cannot be synced because it doesn't have the syncwithvalidationinterfacequeue RPC implemented (required by test_framework.py), so the solution is to copy the block folder to the modern node and start it with -reindex-chainstate. Because of this additional complexity, this testing is best managed in a separate file rather than in the existing migration test files.

In order to know when the node is synced, cannot you just wait for a specific log line? See `busy_wait_for_debug_log`.

w0xlt commented on test/functional/wallet_ancient_migration.py:
```diff
@@ -0,0 +1,292 @@
+
+        # Rescan the blockchain
+        self.log.info("Rescanning blockchain...")
+        self.nodes[1].rescanblockchain()
```
Done. Thanks.

w0xlt commented:
Updated the PR with the same solution used in https://github.com/bitcoin/bitcoin/pull/34240.

deehochberg commented on .github/workflows/ci.yml:
```diff
@@ -505,14 +505,16 @@ jobs:
+          py -3 test/functional/wallet_ancient_migration.py --previous-releases --tmpdir="${RUNNER_TEMP}/test_wallet_ancient_migration"
 
   ci-matrix:
     name: ${{ matrix.name }}
```
Danielle Lawson Hochberg

furszy reviewed (commented):
utACK bf86d56a8696cfb5f4adff1dd2fb99e61d1f655d

maflcko reviewed (commented):
I think the pull description is outdated?

> The node v0.14.3 cannot be synced because it doesn't have the syncwithvalidationinterfacequeue RPC implemented (required by test_framework.py), so the solution is to copy the block folder to the modern node and start it with -reindex-chainstate. Because of this additional complexity, this testing is best managed in a separate file rather than in the existing migration test files.

w0xlt commented on test/functional/wallet_ancient_migration.py:
```diff
@@ -0,0 +1,136 @@
+
+        # Verify all addresses are still owned
+        for addr in all_addresses:
+            assert self.nodes[1].getaddressinfo(addr)['ismine']
```
Thanks. Done in 3a4a7b6859.

brunoerg commented on test/functional/wallet_ancient_migration.py:
```diff
@@ -0,0 +1,146 @@
+        assert_equal(new_wallet_info['format'], 'sqlite')
+        assert_equal(new_wallet_info['descriptors'], True)
+        assert_equal(new_node.getbalance(), old_balance)
+        assert_equal(len(new_node.listtransactions("*", 1000)), len(old_txs))
```
0119f6899ccc56f9ec4576d1a4b401d5c59a29c3: Instead of just comparing the length, it could also compare the content. I think there is a helper for it (`assert_list_txs_equal`).

pablomartin4btc replied:
`assert_list_txs_equal` only strips `parent_descs`. A cross-version comparison (v0.14.3 → modern) would fail due to multiple field differences in `listtransactions` output:

| Field              | v0.14.3 | Modern |
|--------------------|---------|--------|
| `account`          | ✓       | ✗ (removed v0.18) |
| `wtxid`            | ✗       | ✓ (added v22, PR #24198) |
| `mempoolconflicts` | ✗       | ✓ (added recently) |
| `bip125-replaceable`| ✗      | ✓ (deprecated, behind -deprecatedrpc=bip125) |
| `parent_descs`     | ✗       | optional |

A content comparison would need to strip all these fields, not just `parent_descs`. Alternatively, compare only specific fields (txid, amount, category, confirmations).

w0xlt replied:
Thanks, addressed.

I avoided using `assert_list_txs_equal` directly because, as Pablo noted, `listtransactions` output differs between v0.14.3 and current master (`account`, `wtxid`, `mempoolconflicts`, `bip125-replaceable`, `parent_descs`, etc.).

Instead, the test now compares normalized transaction contents using cross-version fields: `txid`, `amount`, `category`, and `confirmations`. This verifies more than just the count while avoiding version-specific fields.

pablomartin4btc commented on test/functional/wallet_ancient_migration.py:
```diff
@@ -0,0 +1,146 @@
+
+        migration_result = new_node.migratewallet("migrated_wallet")
+        assert 'wallet_name' in migration_result
+        assert 'backup_path' in migration_result
```
nit: `assert 'wallet_name' in migration_result` / `assert 'backup_path' in migration_result` — checks key presence only, not values. Perhaps should at least verify `migration_result['wallet_name'] == 'migrated_wallet'`?

w0xlt replied:
Thanks. Done.

pablomartin4btc commented on test/functional/wallet_ancient_migration.py:
```diff
@@ -0,0 +1,146 @@
+
+        # Verify all addresses are still owned
+        for addr in all_addresses:
+            assert new_node.getaddressinfo(addr)['ismine']
```
I think plain `assert` gives no useful failure message...
```suggestion
            assert_equal(new_node.getaddressinfo(addr)['ismine'], True)
```

w0xlt replied:
Thanks. Done.

pablomartin4btc commented on test/functional/wallet_ancient_migration.py:
```diff
@@ -0,0 +1,146 @@
+
+# Wallet version constants from ancient Bitcoin Core (pre-0.15)
+FEATURE_COMPRPUBKEY = 60000  # Non-HD wallet version
+FEATURE_HD = 130000          # HD wallet version
```
nit: `HD wallet version` is a bit imprecise here — `FEATURE_HD_SPLIT = 139900` is also an `HD wallet version`. Since v0.15.0 (PR #9294) introduced the internal keychain split, calling this out as single-chain would make the distinction explicit:
```suggestion
FEATURE_HD = 130000          # Single-chain HD wallet version (pre-v0.15)
```

w0xlt replied:
Thanks. Done.

pablomartin4btc commented on test/functional/wallet_ancient_migration.py:
```diff
@@ -0,0 +1,146 @@
+        old_addresses = [old_node.getnewaddress() for _ in range(4)]
+        old_wallet_info = old_node.getwalletinfo()
+        assert_equal(old_wallet_info['walletversion'], expected_version)
+        assert 'keypoolsize_hd_internal' not in old_wallet_info  # Verify single-chain wallet
```
minor nit: this comment is only accurate for the HD case. For non-HD (expect_hd=False) the field is absent because there's no HD at all, not because it's single-chain...
```suggestion
        assert 'keypoolsize_hd_internal' not in old_wallet_info  # v0.14.3 predates HD split keypool (v0.15+)
```

w0xlt replied:
Thanks. Done.

pablomartin4btc commented on test/functional/wallet_ancient_migration.py:
```diff
@@ -0,0 +1,146 @@
+        # resetting node datadirs between sub-tests.
+        self.add_nodes(self.num_nodes, versions=[140300, None, 140300, None])
+
+    def get_vout_addresses(self, decoded_tx):
```
minor nit: `get_vout_addresses` doesn't use self, I think it could be a module-level function (or `@staticmethod`).

w0xlt replied:
Thanks. Done.

pablomartin4btc reviewed (commented):
ACK 0119f6899ccc56f9ec4576d1a4b401d5c59a29c3

Left a few suggestions.

maflcko commented on test/functional/test_framework/util.py:
```diff
@@ -701,7 +729,7 @@ def dumb_sync_blocks(*, src, dst, height=None):
     for i in range(dst.getblockcount() + 1, height + 1):
         block_hash = src.getblockhash(i)
-        block = src.getblock(blockhash=block_hash, verbosity=0)
+        block = src.getblock(block_hash, False)
```
I don't think it is useful to push unrelated changes to use the deprecated bool alias. This seems backward and is extra churn that may have to be reverted in the future.

w0xlt replied:
v0.14.3’s `getblock` RPC predates numeric `verbosity` and only supports:

`getblock "blockhash" ( verbose )`

where `verbose` is a boolean.

So when `src` is a v0.14.3 node, `src.getblock(blockhash=block_hash, verbosity=0)` fails with `Unknown named parameter verbosity`, and `src.getblock(block_hash, 0)` fails because it expects a boolean.

We can use:

`block = src.getblock(blockhash=block_hash, verbose=False)`

This is accepted by both v0.14.3 and current master, but the legacy boolean parameter is required for the v0.14.3 source node.

maflcko replied:
Ok, interesting. Looks like `verbosity` only made it into 0.15.x. However, named args made it into 0.14.x, according to the release notes:

```
doc/release-notes/release-notes-0.14.0.md:    src/bitcoin-cli -named getblockhash height=0
doc/release-notes/release-notes-0.15.0.md:- The `verbose` argument of `getblock` has been renamed to `verbosity` and now takes an integer from 0 to 2.
```

So maybe:

> `block = src.getblock(block_hash, verbose=False)  # Use boolean verbosity for 0.14.x compatiblity`

or so?

w0xlt replied:
Yes, good idea. Added.

brunoerg commented on test/functional/wallet_ancient_migration.py:
```diff
@@ -0,0 +1,157 @@
+
+        old_balance = old_node.getbalance()
+        old_txs = old_node.listtransactions("*", 1000)
+        assert_greater_than(old_balance, 0)
```
1de6a165efc416f7aff601204df35a70ac7c5cfc: nit: instead of asserting that the balance is greater than 0, we could assert that the returned balance is exactly the expected one. Same for the list of transactions.

w0xlt replied:
Done. Thanks.

brunoerg commented on test/functional/wallet_ancient_migration.py:
```diff
@@ -0,0 +1,157 @@
+        self.log.info("Verifying migration")
+        new_wallet_info = new_node.getwalletinfo()
+        assert_equal(new_wallet_info['format'], 'sqlite')
+        assert_equal(new_wallet_info['descriptors'], True)
```
1de6a165efc416f7aff601204df35a70ac7c5cfc: asserting the txcount would be great as well

```suggestion
        assert_equal(new_wallet_info['descriptors'], True)
        assert_equal(new_wallet_info['txcount'], 105)
```

w0xlt replied:
Done. Thanks.

pablomartin4btc reviewed (commented):
ACK 1de6a165efc416f7aff601204df35a70ac7c5cfc

furszy reviewed (commented):
Code reviewed. What do you think about adding cases for an encrypted wallet too?

pablomartin4btc reviewed (commented):
ACK adf4801fcafadce2f71975582a3fd385264c18d3

Since my last review added both HD and non HD encrypted wallets tests.

pablomartin4btc commented on test/functional/wallet_ancient_migration.py:
```diff
@@ -0,0 +1,200 @@
+        # Test post-migration functionality
+        new_addr = new_node.getnewaddress()
+        if passphrase is not None:
+            new_node.walletpassphrase(passphrase, 60)
```
nit (not blocking): `unlocked_until == 0` (above ~L138) already gives partial coverage — but this would make the encryption assertion airtight...
```suggestion
            assert_raises_rpc_error(-13, "Please enter the wallet passphrase", new_node.sendtoaddress, new_addr, 0.1)
            new_node.walletpassphrase(passphrase, 60)
```

w0xlt replied:
Done. Thanks.

pablomartin4btc commented on test/functional/wallet_ancient_migration.py:
```diff
@@ -0,0 +1,200 @@
+        shutil.copy2(old_wallet_path, migrated_wallet_dir / "wallet.dat")
+
+        migration_kwargs = {} if passphrase is None else {"passphrase": passphrase}
+        migration_result = new_node.migratewallet("migrated_wallet", **migration_kwargs)
```
nit (if you have to re-touch): this can be one-liner, `migratewallet` already checks for passphrase (`RPCArg::Optional::OMITTED`) `.isNull()`, passing `passphrase=None` explicitly serializes to JSON null...
```suggestion
        migration_result = new_node.migratewallet("migrated_wallet", passphrase=passphrase)
```

w0xlt replied:
Done. Thanks.

pablomartin4btc commented on test/functional/wallet_ancient_migration.py:
```diff
@@ -0,0 +1,200 @@
+        migration_kwargs = {} if passphrase is None else {"passphrase": passphrase}
+        migration_result = new_node.migratewallet("migrated_wallet", **migration_kwargs)
+        assert_equal(migration_result['wallet_name'], 'migrated_wallet')
+        assert_greater_than(len(migration_result['backup_path']), 0)
```
nit (not blocking): `migration_result['backup_path']` is only checked for non-empty length — could also verify the backup file actually exists on disk:

```C++
from pathlib import Path
...
assert_greater_than(len(migration_result['backup_path']), 0)
assert Path(migration_result['backup_path']).is_file()

```

so this would catch a real regression (e.g. the backup silently failing to write)...

w0xlt replied:
Done. Thanks.

pablomartin4btc reviewed (commented):
Left a few non blockers nits (only if you have to retouch)...

w0xlt commented:
Rebased and updated:

- Addressed @furszy’s feedback by adding encrypted HD/non-HD variants and explicitly asserting that migration does not rescan.

- Addressed @pablomartin4btc’s follow-up suggestions by verifying locked spending enforcement, always passing the optional passphrase, and confirming the backup is an actual file.

- Added further coverage for single-chain descriptors, complete wallet state, unfunded-address preservation, and absence of unexpected auxiliary wallets.

furszy commented on test/functional/wallet_ancient_migration.py:
```diff
@@ -0,0 +1,229 @@
+        self.skip_if_no_previous_releases()
+
+    def setup_network(self):
+        # Use dedicated old/new node pairs for each migration scenario to avoid
+        # resetting node datadirs between sub-tests.
+        self.add_nodes(self.num_nodes, versions=[140300, None] * 4)
```
The new node supports multi-wallets, so could have 4 old nodes and 1 new one to minimize setup time? Would just need to use a different wallet directory name when you do the copy.

w0xlt replied:
Thanks! Each old node builds an independent chain, so sharing one modern node would require switching its chain between scenarios and introduce cross-scenario state.
Maybe keeping dedicated pairs is preferable here for isolation and simplicity.

furszy replied:
> Each old node builds an independent chain, so sharing one modern node would require switching its chain between scenarios and introduce cross-scenario state.

Or creating the old nodes wallets' txs first, all at once, in the same chain?

np if it is too big of a change at this point. It is better to have this test than to not have it. Can ignore the comment.

furszy reviewed (commented):
utACK ea59f1722097a620de6efd2fedc37b64a105ec96

pablomartin4btc reviewed (commented):
reACK ea59f1722097a620de6efd2fedc37b64a105ec96

Thanks for taking the suggestions!
